File: editor/en/[1071]Greatest Common Divisor of Strings.java
1071
        Greatest Common Divisor of Strings
        2023-02-01 09:29:16

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        int n = str1.length();
        int m = str2.length();
        if (!(str1 + str2).equals(str2 + str1)) {
            return "";
        }
        int max = gcd(n, m);
        return str1.substring(0, max);
    }

    private int gcd(int x, int y) {
        return y == 0 ? x : gcd(y, x % y);
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[1071]Greatest Common Divisor of Strings.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        gcd_length = self.gcd(len(str1), len(str2))
        candidate = str1[:gcd_length]
        if self.is_divisible(str1, candidate) and self.is_divisible(str2, candidate):
            return candidate
        return ""

    def gcd(self, a, b):
        while b:
            a, b = b, a % b
        return a

    def is_divisible(self, s, t):
        if s == t * (len(s) // len(t)):
            return True
        return False

# leetcode submit region end(Prohibit modification and deletion)
